File: tool-list/algorithms/fastsort/fastsort.c
#include <stdio.h>

int my_PARTITION(int A[], int p, int r);

void quicksort(int A[], int p, int r)
{
	if (p < r) {
		int q = my_PARTITION(A, p, r);
		quicksort(A, p, q - 1);
		quicksort(A, q + 1, r);
	}
}

int my_PARTITION(int A[], int p, int r)
{
	int x = A[r];
	int i = p - 1;
	int tmp;

	for (int j = p; j < r; j++) {
		if (A[j] <= x) {
			i++;
			tmp = A[i];
			A[i] = A[j];
			A[j] = tmp;
		}
	}

	tmp = A[i + 1];
	A[i + 1] = A[r];
	A[r] = tmp;
	return i + 1;
}

int main(void)
{
	int arr[8] = {2, 8, 7, 1, 3, 5, 6, 4};

	quicksort(arr, 0, 7);

	for (int i = 0; i < 8; i++)
		printf("%d\n", arr[i]);
	return 0;
}

File: tool-list/algorithms/sorting/fast_sort.py
"""快速排序（Lomuto 分区），含两份等价实现。"""


def fast_sort(nums, start, end):
    def partition(arr, l, r):
        # 以 arr[r] 为基准，s 之前是不大于基准的区域
        key = arr[r]
        s = l - 1

        for i in range(l, r):
            if arr[i] <= key:
                s += 1
                arr[s], arr[i] = arr[i], arr[s]

        arr[s + 1], arr[r] = arr[r], arr[s + 1]
        return s + 1

    if start < end:
        part = partition(nums, start, end)
        fast_sort(nums, start, part - 1)
        fast_sort(nums, part + 1, end)


def quicksort(arr, p, r):
    if p < r:
        q = my_PARTITION(arr, p, r)
        quicksort(arr, p, q - 1)
        quicksort(arr, q + 1, r)


def my_PARTITION(arr, p, r):
    # 以末尾元素为基准；i 始终指向不大于基准区域的末位
    x = arr[r]
    i = p - 1

    for j in range(p, r):
        if arr[j] <= x:
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]
    # 把基准换到 i+1，使其左小右大
    arr[r], arr[i + 1] = arr[i + 1], arr[r]

    return i + 1


def testcase():
    arr = [6, 3, 5, 4, 1, 7, 9, 2, 6, 4, 8, 9, 1, 3, 5, 7, 9, 2, 3]
    fast_sort(arr, 0, len(arr) - 1)
    print(arr)


if __name__ == "__main__":
    testcase()
